### ISSUE TITLE
Proposal: Allow configurable module paths for testify package identification

### ISSUE THREAD

**Issue by hartblanc** [Antonboom/testifylint#288]
## Summary
Currently, `testifylint` identifies `testify` calls by checking for the hardcoded module path `github.com/stretchr/testify` (defined in `internal/testify/const.go`). This proposal suggests adding a configuration option to allow users to specify additional module paths that should be treated as "base" `testify` packages.

## Use Cases
1. **Custom Assertion Wrappers:** Many organizations use forks or internal packages that wrap or embed `testify/assert` to add domain-specific assertions (e.g., custom Proto diffing, or assertions for domain specific types) while preserving the standard `testify` API. 
   - *Example:* A project using `company.com/go/testutils/assert` instead of the direct upstream import.
2. **Internal Forks & Mirrors:** For security or compliance, some companies maintain internal mirrors of dependencies with different import paths (e.g., `git.internal.io/mirrors/testify`).
3. **Legacy Migration:** Projects in the middle of a migration from a custom wrapper back to upstream (or vice versa) may have both paths present and would benefit from linting both simultaneously.

## Proposed Change
Introduce a `base-module-paths` configuration (flag: `--base-module-paths`).

- **Default Value:** `["github.com/stretchr/testify"]`
- **Behavior:** The linter should consider any package matching these paths (or their sub-packages like `/assert`, `/require`, `/suite`) as valid targets for linting.

### Example Configuration (`.golangci.yml`)
```yaml
linters-settings:
  testifylint:
    enable:
      - len
      - expected-actual
    base-module-paths:
      - github.com/stretchr/testify
      - company.com/go/testutils # Custom wrapper package
```

## Implementation Idea
1. **Config Update:** Add `BaseModulePaths []string` to `internal/config/config.go`.
2. **Analyzer logic:** Refactor `analyzer/analyzer.go`'s file pre-filtering to check imports against the configured list rather than the `testify.ModulePath` constant.
3. **CallMeta logic:** Update `internal/checkers/call_meta.go` to validate the `initiatorPkg` against the configured paths.

## Backward Compatibility
The default behavior remains unchanged, as the list will default to the current hardcoded upstream path.

---
I am happy to work on a Pull Request for this if the maintainers are open to the idea!

**Comment by hartblanc** [comment]
Hey @ccoVeille,

Thanks for taking a look. Any feedback on this idea?

**Comment by Antonboom** [comment]
hi, @hartblanc 

usually `testify` forks just fix all issues of `testify` from which my linter guards 😆 

do we really need it?

**Comment by hartblanc** [comment]
Hi @Antonboom,

In my organisation our internal fork is essentially the same as testify just with some additional org-specific assertions. We have extensive use (100ks-millions of calls) of the sub-optimal APIs throughout our codebase. As is the case for many users of the standard testify API, changing the API, updating all the call-sites, and maintaining a more divergent fork is considered to be too risky and time-intensive.

So instead we are hoping to detect and automatically fix common issues using this linter.

**Comment by adambaratz** [comment]
I’m in the same situation as @hartblanc. We landed on the same approach to using testify. This linter would be a big help for standardizing a number of things, but without this change we’d have to fork it. With the adoption of LLM coding tools, it’s become important to us to have more linters to enforce consistency in our repo.

**Comment by Antonboom** [comment]
and you not only fork it, but the change module path (in `go.mod`) too? yep?

**Comment by hartblanc** [comment]
> change module path (in go.mod) too? yep?

Yep, we have a different module path.

> and you not only fork it

Apologies, I was being a little imprecise.

In our particular case, we don't strictly have a fork. We have a library within our monorepo (and therefore within our go module) where we have our own types that embed the equivalent testify types (e.g. `suite.Suite`, `assert.Assertions`, `require.Assertions`) and add some additional methods.

But for all intents and purposes, yes.
It is essentially the same problem as if we had forked and changed the module path in go.mod.

**Comment by Antonboom** [comment]
Can you, please, share the small example with dir tree, package and module naming and your own types definition and some methods (I'm trying to understand the "complexity" of the required config)

**Comment by adambaratz** [comment]
Speaking for my own team, we have in-house packages: `testing/assert, `testing/mock`, and `testing/require` (nested within a relevant place in a monorepo).

They contain files like this:

```go
package assert

import (
	"github.com/stretchr/testify/assert"
)

var AnError = assert.AnError
var Contains = assert.Contains
var ElementsMatch = assert.ElementsMatch
// etc.
```

Then there are additional files in those packages for the functionality we've layered on. The idea was to have these packages act like a testify superset so we don't have to think about which functionality comes from testify and which comes from some other package in our repo. Throughout our repo, there are `assert.Equal`/similar calls that look just like testify, as well as some asserts that are specific to our repo.

The proposal at the top of this thread should work just fine for us. If we could point ModulePath at an internal path instead of "github.com/stretchr/testify", I would expect the linter to otherwise be a great help for us.

**Comment by hartblanc** [comment]
Ours is very similar. We wrap the testify API a little more verbosely though. see an example below.
We also make use of the assert.Assertions, require.Assertions objects.
I was able to patch testifylint with the proposed changes and it worked well in our case.

```go
package assert

import (
	"github.com/stretchr/testify/assert"
)

type Assertions struct {
	t assert.TestingT
	*assert.Assertions
}

// Nil delegates to Nil(t, ...)
func (a *Assertions) Nil(object interface{}, msgAndArgs ...interface{}) bool {
	return Nil(a.t, object, msgAndArgs...)
}

// ...


func Nil(t assert.TestingT, object interface{}, msgAndArgs ...interface{}) bool {
	return assert.Nil(t, object, msgAndArgs...)
}
```

**Comment by Antonboom** [comment]
Thanks a lot

Looks like not other packages support, but support of wrappings and nesting of `stretchr/testify`, e.g. how format linters do 🤔
